### ISSUE TITLE
[css-overflow-5]: allowing tab to access non-current items causes mismatch in forward vs. backward tab order

### ISSUE THREAD

**Issue by smhigley** [w3c/csswg-drafts#12270]
This issue can be observed on the Cards example in the gallery: https://chrome.dev/carousel/horizontal/cards/

If the carousel is set to any marker but the first, the behavior tabbing forwards through the carousel does not match the behavior tabbing backwards.

The solution I would lean towards would be to ensure that tabbing does not reach any carousel item that is not fully in the viewport, so that tab alone can never trigger slide changes.

**Comment by smhigley** [comment]
cc/ @flackr (because he asked me to, not just randomly tagging folks :D)

**Comment by flackr** [comment]
This should be fixed by #12122.

**Comment by smhigley** [comment]
The original behavior is fixed, but that fix introduced a new problem (or at least I think it wasn't there before): now when I tab into the slide content and use arrow keys to scroll, my keyboard focus is lost if the focused-on slide moves offscreen.

I do think the behavior of offscreen slides being inert is a good one, but there would need to either be focus handling between slides when the currently-focused content goes offscreen, or the arrow-to-scroll behavior needs to be reworked.

**Comment by flackr** [comment]
We already have a special case for arrow scrolling where if you don't have a focused element it targets the last mouse down node - this is why you can click in a scroller on a non-focusable element and still use arrow keys to scroll that scroller. This also retargets to the nearest still attached ancestor when that element is removed. I think the same should apply when the focused element is removed or becomes no longer focusable.

Similarly when the focused element is removed from the dom or becomes unfocusable I think we should set the [sequential focus navigation starting point](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation-starting-point) such that focus will start from the location that just had focus.

**Comment by smhigley** [comment]
If I've understood you correctly, that doesn't actually solve the focus issue -- the problem is that focus should never be lost while scrolling regardless of what happens with focus navigation start point. Focus loss affects screen reader cursor as well, not just keyboard operability, which is why it matters.

There are essentially three reasonable and accessible keyboard behaviors for a control like this:
1. It acts like a normal scrolling region where offscreen content is not inert or truly hidden, in which case focus remains wherever the user put it while scrolling. This works best for link behavior situations like the table of contents.
2. Arrow keys move the user's focus between slides, and slides have semantics that make this behavior predictable and understandable (not tabpanel). This would effectively make the set of slides into a composite widget (examples include arrowing between tabs in a tablist, menuitems in a menu, articles in a feed, or I'd even consider listitems in a list here even if it's not 100% kosher yet). This would need much more detailed focus management, since arrowing while focus is on the inner content of a slide shouldn't scroll and users need an easy way to get back to arrowing between slides after moving their focus into the content.
3. Slides don't scroll at all (for mouse or keyboard). The only way to change slides is to use the scroll markers or scroll buttons.

Let me know if that makes sense / if I've missed anything!

**Comment by flackr** [comment]
> If I've understood you correctly, that doesn't actually solve the focus issue -- the problem is that focus should never be lost while scrolling regardless of what happens with focus navigation start point. Focus loss affects screen reader cursor as well, not just keyboard operability, which is why it matters.

I agree we want this to work as expected for screen readers as well. What are developers expected to do when you activate a control that removes the currently focused element (e.g. a delete button)? We could move focus to the newly active scroll marker if using tabs mode and focus was in the previous slide which might match expectations.

> There are essentially three reasonable and accessible keyboard behaviors for a control like this:
> 
> 1. It acts like a normal scrolling region where offscreen content is not inert or truly hidden, in which case focus remains wherever the user put it while scrolling. This works best for link behavior situations like the table of contents.

This is the default behavior if the author doesn't change from the links mode: https://drafts.csswg.org/css-overflow-5/#valdef-scroll-marker-group-links

> 2. Arrow keys move the user's focus between slides, and slides have semantics that make this behavior predictable and understandable (not tabpanel). This would effectively make the set of slides into a composite widget (examples include arrowing between tabs in a tablist, menuitems in a menu, articles in a feed, or I'd even consider listitems in a list here even if it's not 100% kosher yet).

The APG pattern for slides with scroll markers suggests that they are tabpanels, https://www.w3.org/WAI/ARIA/apg/patterns/carousel/examples/carousel-2-tablist/. While the APG example does not allow using the arrow keys to move between slides, many developers choose to have this behavior. As mentioned in the next point, developers can choose when they want the panels / slides to be user scrollable by whether they set `overflow: hidden` or `overflow: auto`.

>    This would need much more detailed focus management, since arrowing while focus is on the inner content of a slide shouldn't scroll and users need an easy way to get back to arrowing between slides after moving their focus into the content.

This is standard scrolling behavior. If the focused element does not preventDefault or otherwise handle the keypress, then it is allowed to scroll the nearest ancestor scroller that can scroll in this direction. if the inner content does handle the arrow key (e.g. text input, inner scroller, developer handled), then it will prevent scrolling an ancestor. Being able to scroll the ancestor scroller while focus is on inner content is not new to this feature.

> 3. Slides don't scroll at all (for mouse or keyboard). The only way to change slides is to use the scroll markers or scroll buttons.> 
> Let me know if that makes sense / if I've missed anything!

This is an author decision whether the slides are user scrollable. If they use `overflow: hidden` then the slides can only be navigated by the scroll buttons or scroll markers.

**Comment by smhigley** [comment]
Unfortunately, I don't think having focus jump anywhere on scroll is a behavior that would be predictable or expected. Either arrow keys scroll and focus doesn't move, or arrow keys move focus between elements with semantics that support that behavior (where the focus movement itself can then cause scrolling). Any other behavior would be an accessibility issue -- imagine someone using a screen reader, bumping their trackpad or mouse, and having focus randomly jump around.

This is a fundamentally different behavior from focus moving after a click-type interaction. Certain actions can prime the user to expect or at least not be surprised by a change of context, and others (scroll in this case) do not.

Everything else in my comment was me laying out more specific options and details of what those two behaviors could look like in the context of a carousel. I think you might be mixing those up a bit, and picking and choosing examples of semantics and keyboard behavior separately to try to find a basis for the current approach. Unfortunately, the semantics and behavior are interdependent, so if you want to e.g. follow the APG carousel tabpanel approach, the keyboard behavior needs to stay consistent with those semantics.

To address a couple specifics:

> The APG pattern for slides with scroll markers suggests that they are tabpanels, https://www.w3.org/WAI/ARIA/apg/patterns/carousel/examples/carousel-2-tablist/. While the APG example does not allow using the arrow keys to move between slides, many developers choose to have this behavior.

Tabpanels are fine as a semantic if you can't lose focus by scrolling between them without interacting with the tablist. Similarly, being able to move through carousel slides without interacting with the scroll markers/buttons is a common behavior devs/designers want, as you mentioned. That also isn't a problem in itself (I've made a number of these myself), but it requires different semantics and keyboard behaviors to be accessible. Allowing people to combine the two out of the box is the problem. If the browser provides the behavior of navigating between slides without interacting with the scroll markers/buttons out of the box, then it should provide semantics and keyboard behavior to match at the same time. The APG carousel + tab semantics example is no longer something that can be referenced as soon as this scrolling + offscreen inert behavior is added.

To make sure I'm not misinterpreted -- I think offscreen inert for carousel slides is a good thing, since not needing to navigate through the entire carousel before continuing is a primary UX feature. I also have made enough composite-widget-with-slide-arrowing carousels that I understand the desire to provide that feature. The only issue is combining standard scrolling with tabpanel semantics, no composite widget arrow key behavior, and focus loss.

> Being able to scroll the ancestor scroller while focus is on inner content is not new to this feature.

Similarly, scrolling is not new, but elements losing focus on scroll _is_ new, and that is the root of the problem.

> This is standard scrolling behavior. If the focused element does not preventDefault or otherwise handle the keypress, then it is allowed to scroll the nearest ancestor scroller that can scroll in this direction.

I was describing manged focus composite widgets in that section, not scrolling. The crux of what I was trying to describe is that you do _not_ allow standard keyboard scrolling within a managed focus widget like a listbox or menu. Arrow keys move focus between top-level items, which is how you avoid

I hope that helps clarify the issue, and some of the possible solution space. I can also make a short video of what happens to a screen reader cursor on a trackpad bump + focus loss if that would help.

**Comment by bramus** [comment]
> Unfortunately, I don't think having focus jump anywhere on scroll is a behavior that would be predictable or expected. 

IUC, the thing that @flackr proposed in [this comment](https://github.com/w3c/csswg-drafts/issues/12270#issuecomment-5622724236) does not have focus jump around:

> Similarly when the focused element is removed from the dom or becomes unfocusable I think we should set the [sequential focus navigation starting point](https://html.spec.whatwg.org/multipage/interaction.html#sequential-focus-navigation-starting-point) such that focus will start from the location that just had focus.

Practically, I am trying things out in Chrome 154+ with https://codepen.io/editor/bramus/pen/01a0571c-931b-743f-a352-476408feeec9?orientation=left&panel=files&show=preview as a demo. Try this scenario:

1. Use the form to set the mode to `tabs`
2. Hit TAB to focus the scroll markers, with “Slide 1” now focused
3. Hit the right arrow to move focus to slide 2 in the scroll markers bar (and to also see the carousel move to slide 2)
4. Hit TAB to focus the “Interactive Element 2” button
5. Hit the left arrow (and see the scroller scroll to the first slide + see the scroll markers update; yet “Interactive Element 2” remains focused!)
6. Hit SHIFT+TAB

After doing step 6, the focus currently jumps to the “Tabs option”, which is definitely wrong.

https://github.com/user-attachments/assets/3c1cfc77-9a5f-4f13-8013-4852b2a358db

I would have expected either:

- The focus to jump to the “Interactive Element 1” button after step 5 _(but that is a no-go IUC)_
- The focus to jump to the “Interactive Element 1” button after step 6 _(which is that “sequential focus navigation starting point”?)_

**Comment by smhigley** [comment]
@bramus thanks for weighing in, testing, and finding that odd glitch! Definitely seems like some sort of bug around when partially-offscreen content gets removed from the tab order, though especially odd that focus remained on the button.

Re: focus navigation starting point -- I agree that it should be set when focus is lost from an element within a carousel, similar to any other similar focused-element-removal scenario. However, this is entirely unrelated to the issue I was talking about, which is why I didn't address that specific suggestion in my earlier comment.

The accessibility issue I'm talking about is this one:
> imagine someone using a screen reader, bumping their trackpad or mouse, and having focus randomly jump around.

The sequential focus navigation starting point is a really nice browser feature, but in no way does it remove the need for focus management when focus is lost. I actually wrote a post some time ago that explains why in more detail (the AT behavior part starts from the "How SFNSP works today" heading): https://sarahmhigley.com/writing/focus-navigation-start-point/

**Comment by bramus** [comment]
> The accessibility issue I'm talking about is this one:
> 
> > imagine someone using a screen reader, bumping their trackpad or mouse, and having focus randomly jump around.

I get the problem, but I’d argue that simply _“bumping”_ the trackpad of mouse doesn’t perform a scroll, but requires a bit more to occur at the same time _(e.g. scroll on trackpad = typically needs two fingers instead of just one)_. So perhaps we’re quite safe on that one.

Also, doesn’t the same apply to accidentally activating a button that opens a modal dialog? If you erroneously activate the button then focus will jump around, which is done regardless of that trigger getting triggered by accident or not.

**Comment by smhigley** [comment]
I've seen a number of unintentional scroll jumps in user studies, which would indicate that it is an issue one way or another 😅. Screen reader users can also very easily be in a different mode than they expect, and accidentally use arrow keys to scroll (i.e. in forms mode) while trying to use them to move the cursor (in virtual cursor/browse mode).

That's also not even touching on the partially sighted screen reader users who might scroll intentionally, then find themselves way back at the top of the page when focus is lost.

This is not the same as erroneously activating a button, and even buttons are required to not jump focus when clicked in certain cases (ref: [WCAG's On Input criterion](https://w3c.github.io/wcag/understanding/on-input.html))

I'm a bit surprised that this angle is getting any pushback at all -- generally I'd consider "focus shouldn't get lost on scroll" to be a basic, uncontroversial accessibility requirement. There's a good argument this also violates WCAG's On Input requirement as well, though generally I prefer not to approach accessibility issues by litigating WCAG.

**Comment by aardrian** [comment]
> I get the problem, but I’d argue that simply “bumping” the trackpad of mouse doesn’t perform a scroll, but requires a bit more to occur at the same time (e.g. scroll on trackpad = typically needs two fingers instead of just one). So perhaps we’re quite safe on that one.

FWIW, I've sat through lots of testing sessions with SR users where the page very much scrolls around without them meaning to, and it's not from bumping the pointer. Sometimes it's a function of messing up a navigation action, sometimes it's a novice user SR who still struggles with their kit, sometimes it's a sighted user (and maybe novice) user trying to mouse around and pitting the two interaction modes against one another, sometimes it's a mobility-impaired SR user (or situationally-mobility-impaired SR user), sometimes it's a mis-tap because of gloves or water on the screen or leftover taffy on their hands for some reason (we really need to check for taffy smugglers in these sessions), and so on. The point is that it's _not_ uncommon.

So consider a user who has taken an unintentional action and who has to recover from that and _then_ consider how they handle that recovery when the UI has changed more than a typical mis-scroll happens.

And while Sarah may not want to litigate WCAG, I have no issue doing so. Especially since I've spent years cleaning up when it's not considered at the start. Hence my [public ask](https://adrianroselli.com/2025/05/my-request-to-google-on-accessibility.html#Request) to anyone working on new features for the web platform:

> Please, if your team cannot explain how _the thing_ satisfies all WCAG Success Criteria at Level AA, then don’t release the thing.
